sunscreen/src/types/zkp/rns_polynomial.rs

```diff
@@ -0,0 +1,156 @@
+use sunscreen_compiler_macros::TypeName;
+use sunscreen_runtime::ZkpProgramInputTrait;
+use sunscreen_zkp_backend::BigInt;
+
+use crate::{
+    types::zkp::{Coerce, ProgramNode},
+    with_zkp_ctx,
+    zkp::ZkpContextOps,
+};
+
+use super::{AddVar, NativeField, NumFieldElements, ToNativeFields, ZkpType};
+
+use crate as sunscreen;
+
+/**
+ * A polynomial in Z_q[X]/(X^N+1), up to degree N-1. `q` is the
+ * coefficient modulus and is the product of R factors. Each coefficient is decomposed
+ * into R residues (i.e. RNS form).
+ *
+ * # Remarks
+ * Operations (e.g. add, mul, etc) don't automatically reduce modulo q. This enables
+ * program authors to batch multiple operations before reduction.
+ */
+#[derive(Debug, Clone, TypeName)]
+pub struct RnsRingPolynomial<const N: usize, const R: usize> {
+    data: Box<[[NativeField; N]; R]>,
+}
+
+impl<T, const N: usize, const R: usize> From<[[T; N]; R]> for RnsRingPolynomial<N, R>
+where
+    T: Into<NativeField> + std::fmt::Debug,
+{
+    fn from(x: [[T; N]; R]) -> Self {
+        Self {
+            data: Box::new(
+                x.into_iter()
+                    .map(|x| {
+                        x.into_iter()
+                            .map(|x| x.into())
+                            .collect::<Vec<NativeField>>()
+                            .try_into()
+                            .unwrap()
+                    })
+                    .collect::<Vec<[NativeField; N]>>()
+                    .try_into()
+                    .unwrap(),
```

Comment:
Is this code also temporary? If not, might be better to impl TryFrom/TryInto and unwrap at call site, unless we know this is infallible.

Reply:
TIL, `array.map([T; N]) -> [U; N]`. That cleans the iteration nonsense up.